src/coreclr/jit/assertionprop.cpp

```diff
@@ -5591,6 +5591,141 @@ void Compiler::optImpliedAssertions(AssertionIndex assertionIndex, ASSERT_TP& ac
     }
 }
 
+//------------------------------------------------------------------------
+// optCreateJumpTableImpliedAssertions: Create assertions for the switch statement
+//     for each of its jump targets.
+//
+// Arguments:
+//     switchBb - The switch statement block.
+//
+// Returns:
+//     true if any modifications were made, false otherwise.
+//
+bool Compiler::optCreateJumpTableImpliedAssertions(BasicBlock* switchBb)
+{
+    assert(!optLocalAssertionProp);
+    assert(switchBb->KindIs(BBJ_SWITCH));
+    assert(switchBb->lastStmt() != nullptr);
+    bool modified = false;
+
+    GenTree* switchTree = switchBb->lastStmt()->GetRootNode()->gtEffectiveVal();
+    assert(switchTree->OperIs(GT_SWITCH));
+
+    // bbsCount is uint32_t, but it's unlikely to be more than INT32_MAX.
+    noway_assert(switchBb->GetSwitchTargets()->bbsCount <= INT32_MAX);
+
+    ValueNum opVN = optConservativeNormalVN(switchTree->gtGetOp1());
+    if (opVN == ValueNumStore::NoVN)
+    {
+        return modified;
+    }
+
+    if (vnStore->TypeOfVN(opVN) != TYP_INT)
+    {
+        // Should probably be an assert instead - GT_SWITCH is expected to be TYP_INT.
+        return modified;
+    }
+
+    // Typically, the switch value is ADD(X, -cns), so we actually want to create the assertions for X
+    int offset = 0;
+    vnStore->PeelOffsetsI32(&opVN, &offset);
+
+    int        jumpCount  = static_cast<int>(switchBb->GetSwitchTargets()->bbsCount);
+    FlowEdge** jumpTable  = switchBb->GetSwitchTargets()->bbsDstTab;
+    bool       hasDefault = switchBb->GetSwitchTargets()->bbsHasDefault;
+
+    for (int jmpTargetIdx = 0; jmpTargetIdx < jumpCount; jmpTargetIdx++)
+    {
+        // The value for each target is jmpTargetIdx - offset.
+        if (CheckedOps::SubOverflows(jmpTargetIdx, offset, false))
+        {
+            continue;
+        }
+        int value = jmpTargetIdx - offset;
+
+        // We can only make "X == caseValue" assertions for blocks with a single edge from the switch.
+        BasicBlock* target = jumpTable[jmpTargetIdx]->getDestinationBlock();
+        if ((target->GetUniquePred(this) != switchBb) || (fgGetPredForBlock(target, switchBb)->getDupCount() > 1) ||
+            !BasicBlock::sameEHRegion(target, switchBb))
```

Comment:
Do we need to bail if not same EH region? If the switch arm is a try entry still seems fine to add the assertion.

Reply:
hm.. I think you're right, relaxed